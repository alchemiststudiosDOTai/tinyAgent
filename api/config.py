from tinyagent.config.config import load_config


def get_api_config():
    """Get API-specific configuration using existing config loader"""
    config = load_config()
    return {
        "app_name": "AI Agent API",
        "enable_docs": config.get("api", {}).get("enable_docs", True),
        "cors_origins": config.get("api", {}).get("cors_origins", ["*"]),
        "port": config.get("api", {}).get("port", 9000)
    }
