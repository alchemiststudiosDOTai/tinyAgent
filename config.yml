# tinyAgent Configuration
# Model configuration
#base_url: "http://localhost:1234/v1"
base_url: "https://openrouter.ai/api/v1"

model:
  # Default model to use
  #default: "deepseek/deepseek-r1"
  #default: "mistral-small-24b-instruct-2501"
  #default: "anthropic/claude-3.7-sonnet"
  default: "deepseek/deepseek-chat"
  # use claude 3.7 for best resutls, v3 is best value while using tools but 3.7 is the SOTA to use if you can afford it, r1 is best for general use
  
# Parsing configuration
parsing:
  # When true, only strict JSON parsing is used
  strict_json: false
  # Fallback parsers (only used when strict_json is false)
  fallback_parsers:
    # Enable template-style parsing as first fallback
    template: true
    # Enable regex-style parsing as last resort
    regex: true

# Retry configuration with warming and escalation
retries:
  # Maximum retry attempts
  max_attempts: 3  
  # Temperature control for creativity
  temperature:
    initial: 0.2
    increment: 0.4
    max: 1
  # Model escalation sequence
  model_escalation:
    enabled: true
    sequence:
      - "deepseek/deepseek-chat"      # Start with basic
      - "anthropic/claude-3.5-sonnet"  # Escalate to better
      - "anthropic/claude-3.7-sonnet"  # Final escalation to best

# Rate limiting configuration
rate_limits:
  # Global rate limit (applies to all tools by default)
  global: 2
  # Tool-specific rate limits (override global)
  tools:
    file_hunter: 5
    ripgrep: 10
    chat: -1  # No limit for chat tool

# Dynamic agent configuration
dynamic_agents:
  # Whether to allow creating new tools without asking
  allow_new_tools_by_default: false
  # Whether to use existing tools if possible before creating new ones
  prioritize_existing_tools: true
  # Maximum number of dynamic agents to keep in memory
  max_agents: 10
  # Model to use for dynamic agent operations (defaults to model.default if not specified)
  model: "deepseek/deepseek-chat"
# ElderBrain configuration
elder_brain:
  # Enable ElderBrain by default for all tasks, this is a beta feature and may not work as expected 
  # this is mostly for agents that need to handle complex tasks that may require multiple steps frequently
  default_enabled: false
  # Verbosity level for ElderBrain output (higher value = more detailed output)
  verbosity: 2
  # Model to use for ElderBrain operations (defaults to model.default if not specified)
  model: null
# Code execution security configuration
code_execution:
  # Allow potentially dangerous operations (file operations, etc)
  allow_dangerous_operations: true  # Set to false to enforce strict security
  # Optional: more granular control
  allowed_operations:
    file_operations: true
    os_operations: false
    imports: ["os", "sys"]  # Additional allowed imports beyond the defaults
# File operations configuration
file_operations:
  # Default output directory for file operations
  output_dir: "tinyAgent_output"
  # List of allowed file extensions (empty list means all extensions allowed)
  allowed_extensions:
    - ".txt"
    - ".json"
    - ".yaml"
    - ".md"
    - ".py"
    - ".log"
  # Maximum file size in bytes (10MB default)
  max_file_size: 10485760
  # Whether to create subdirectories automatically
  create_subdirs: true
  # Whether to allow overwriting existing files
  allow_overwrite: true
# Proxy configuration
proxy:
  # Master switch for proxy functionality
  # When false, all proxy features are disabled regardless of other settings
  # When true, the proxy will be used according to the configuration below
  enabled: False

  # Proxy URL template with placeholders
  # Format: http://customer-{username}-cc-{country}:{password}@pr.oxylabs.io:7777
  # The %s placeholders will be replaced in order with:
  # 1. username (from TINYAGENT_PROXY_USERNAME env var)
  # 2. country code (from TINYAGENT_PROXY_COUNTRY env var)
  # 3. password (from TINYAGENT_PROXY_PASSWORD env var)
  # basd on your proxy provider the url format can be diffrent, these confgis are for oxy
  url: "http://customer-%s-cc-%s:%s@pr.oxylabs.io:7777"
  # Authentication credentials loaded from environment variables
  # These values will be inserted into the URL template above
  username: ${TINYAGENT_PROXY_USERNAME}
  password: ${TINYAGENT_PROXY_PASSWORD}
  country: ${TINYAGENT_PROXY_COUNTRY}
  
logging:
  level: INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
  format: "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
  handlers: ["console"] 
