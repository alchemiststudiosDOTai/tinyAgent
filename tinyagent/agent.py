"""
tinyagent.agent
Minimal, typed ReAct agent (Reason + Act) with JSON-tool calling.

Public surface
--------------
ReactAgent  – class
"""

from __future__ import annotations

import json
import os
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any, Final

from openai import OpenAI

from .prompt import BAD_JSON, SYSTEM  # prompt.py holds the two template strings
from .tools import Tool, get_registry  # our Tool wrapper and registry

__all__ = ["ReactAgent"]

# ---------------------------------------------------------------------------
MAX_STEPS: Final = 10
TEMP_STEP: Final = 0.2


class StepLimitReached(RuntimeError):
    """Raised when no answer is produced within MAX_STEPS."""


@dataclass(kw_only=True)
class ReactAgent:
    """
    A lightweight ReAct loop.

    Parameters
    ----------
    tools
        Sequence of Tool objects (typically produced by @tool decorator).
    model
        Model name (OpenAI or OpenRouter format). Default ``gpt-4o-mini``.
        Examples: ``gpt-4``, ``anthropic/claude-3.5-haiku``, ``meta-llama/llama-3.2-3b-instruct``
    api_key
        Optional OpenAI key; falls back to ``OPENAI_API_KEY`` env var.
    """

    tools: Sequence[Tool]
    model: str = "gpt-4o-mini"
    api_key: str | None = None

    # ------------------------------------------------------------------
    def __post_init__(self) -> None:
        if not self.tools:
            raise ValueError("ReactAgent requires at least one tool.")

        # Get the registry to look up Tool objects for functions
        registry = get_registry()

        # Build tool map, handling both Tool objects and functions
        self._tool_map: dict[str, Tool] = {}
        for item in self.tools:
            if isinstance(item, Tool):
                self._tool_map[item.name] = item
            elif callable(item) and item.__name__ in registry:
                # Function decorated with @tool
                self._tool_map[item.__name__] = registry[item.__name__]
            else:
                raise ValueError(f"Invalid tool: {item}")

        # Initialize OpenAI client
        api_key = self.api_key or os.getenv("OPENAI_API_KEY")
        base_url = os.getenv("OPENAI_BASE_URL")
        self.client = OpenAI(api_key=api_key, base_url=base_url)

        # Render immutable system prompt once
        self._system_prompt: str = SYSTEM.format(
            tools="\n".join(
                f"- {t.name}: {t.doc or '<no description>'} | args={t.signature}"
                for t in self._tool_map.values()
            )
        )

    # ------------------------------------------------------------------
    def run(self, question: str, *, max_steps: int = MAX_STEPS) -> str:
        """
        Execute the loop and return the final answer or raise StepLimitReached.
        """
        messages: list[dict[str, str]] = [
            {"role": "system", "content": self._system_prompt},
            {"role": "user", "content": question},
        ]
        temperature = 0.0

        for _ in range(max_steps):
            assistant_reply = self._chat(messages, temperature)
            payload = self._try_parse_json(assistant_reply)

            # If JSON malformed → ask model to fix
            if payload is None:
                messages += [
                    {"role": "assistant", "content": assistant_reply},
                    {"role": "user", "content": BAD_JSON},
                ]
                temperature += TEMP_STEP
                continue

            # Final answer path
            if "answer" in payload:
                return str(payload["answer"])

            # Tool invocation path
            name = payload.get("tool")
            args = payload.get("arguments", {}) or {}
            if name not in self._tool_map:
                return f"Unknown tool '{name}'."

            result = self._safe_tool(name, args)
            messages += [
                {"role": "assistant", "content": assistant_reply},
                {"role": "user", "content": f"Tool '{name}' returned: {result}"},
            ]

        raise StepLimitReached("Exceeded max ReAct steps without an answer.")

    # ------------------------------------------------------------------
    def _chat(self, messages: list[dict[str, str]], temperature: float) -> str:
        """Single LLM call; OpenAI-compatible."""
        response = self.client.chat.completions.create(
            model=self.model,
            messages=messages,
            temperature=temperature,
        )
        return response.choices[0].message.content.strip()

    @staticmethod
    def _try_parse_json(text: str) -> dict[str, Any] | None:
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            return None

    def _safe_tool(self, name: str, args: dict[str, Any]) -> str:
        try:
            return self._tool_map[name].run(args)
        except Exception as exc:  # pragma: no cover
            return f"Error in tool '{name}': {exc}"
