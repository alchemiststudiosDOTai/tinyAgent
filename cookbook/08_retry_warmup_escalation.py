"""
Retry Warmup Escalation (RWE) Demo

This example demonstrates the enhanced retry system with:
1. Temperature warming
2. Model escalation
3. Detailed retry tracking

The agent will progressively:
- Increase temperature for creativity
- Escalate to more capable models
- Track and report all retry attempts
"""

import os
import sys
import json
from typing import Dict, Any
import logging

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from core.agent import Agent
from core.config import load_config

def create_demo_config() -> Dict[str, Any]:
    """Create a demo configuration with retry settings."""
    base_config = load_config()
    
    # Add retry-specific configuration
    base_config.update({
        'retries': {
            'max_attempts': 3,
            'temperature': {
                'initial': 0.2,
                'increment': 0.2,
                'max': 0.8
            },
            'model_escalation': {
                'enabled': True,
                'sequence': [
                    "deepseek/deepseek-chat",      # Basic model
                    "anthropic/claude-3.5-sonnet",  # Better model
                    "anthropic/claude-3.7-sonnet"   # Best model
                ]
            }
        }
    })
    
    return base_config

class FailingAnalyzer:
    """A tool that fails the first two times to demonstrate retry behavior."""
    
    def __init__(self):
        self.attempts = 0
        
    def analyze(self, text: str) -> str:
        self.attempts += 1
        
        if self.attempts == 1:
            raise ValueError("First attempt: Basic model failed to understand complex quantum mechanics")
        elif self.attempts == 2:
            raise ValueError("Second attempt: Better model struggled with parallel universe concepts")
        else:
            return f"Final Analysis (Best Model): Successfully analyzed {text}"

def demo_retry_sequence(agent: Agent) -> None:
    """Run a sequence that demonstrates all retry stages."""
    
    # Create our failing analyzer
    analyzer = FailingAnalyzer()
    
    # Register the tool with the agent
    agent.create_tool(
        name="analyze",
        description="Analyze complex scientific concepts",
        func=analyzer.analyze,
        parameters={"text": "The text to analyze"}
    )
    
    try:
        print("\n=== Starting Analysis ===")
        result = agent.run(
            "Please analyze the quantum entanglement implications of "
            "using a neural network to predict weather patterns in "
            "a fifth-dimensional space while considering the butterfly effect "
            "in parallel universes where time flows backwards."
        )
        print("\n=== Final Success ===")
        print(f"Result: {result}")
        
    except Exception as e:
        print("\n=== Analysis Failed ===")
        print(f"Final Error: {str(e)}")
        if hasattr(e, 'history'):
            print("\n=== Retry History ===")
            for attempt in e.history:
                print(f"\nAttempt {attempt['attempt']}:")
                print(f"Model: {attempt.get('model', 'unknown')}")
                print(f"Temperature: {attempt.get('temperature', 'unknown')}")
                print(f"Error: {attempt.get('error', 'unknown')}")

def main():
    # Configure logging
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )
    
    # Load demo config
    config = create_demo_config()
    
    # Create agent with demo config
    agent = Agent(config=config)
    
    # Print demo header
    print("\n=== RWE Demo ===")
    print("This will demonstrate the retry system with:")
    print("1. Temperature warming (0.2 → 0.4 → 0.6)")
    print("2. Model escalation (basic → better → best)")
    print("3. Detailed retry tracking")
    
    # Run the demo
    demo_retry_sequence(agent)

if __name__ == "__main__":
    main() 