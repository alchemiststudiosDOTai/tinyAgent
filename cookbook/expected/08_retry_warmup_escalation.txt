Using model: deepseek/deepseek-r1
Max steps per research phase: 2
Debug level: 0
Configuration source: defaults

Research phases:
  - gathering: 2 steps
  - analysis: 2 steps
  - synthesis: 2 steps
Initializing custom text browser...
=== CONFIGURING PROXY ===
Proxy disabled in configuration
Proxy disabled for browser
Initializing file manager...
Research output directory: tinyAgent_output/research
Initializing tools registry...
Registered 8 gathering tools, 5 analysis tools, and 3 synthesis tools

Initialization complete

Using model: deepseek/deepseek-r1
Max steps per research phase: 2
Debug level: 0
Rate limit delay: 2.0s
Max retries: 3
Retry delay: 1.0s
Proxy configured: False

=== RWE Demo ===
This will demonstrate the retry system with:
1. Temperature warming (0.2 → 0.4 → 0.6)
2. Model escalation (basic → better → best)
3. Detailed retry tracking

=== Starting Analysis ===

[OpenRouterRequest] Structured outputs DISABLED. Building standard payload without schema.

[OpenRouterRequest] Structured outputs DISABLED. Building standard payload without schema.

[OpenRouterRequest] Structured outputs DISABLED. Building standard payload without schema.

=== Final Success ===
Result: Final Analysis (Best Model): Successfully analyzed The quantum entanglement implications of using neural networks for weather prediction in fifth-dimensional space with reverse-time parallel universes involves multiple complex theoretical concepts. This combines quantum mechanics principles like entanglement with higher dimensional mathematics, chaos theory (butterfly effect), and theoretical physics concepts of parallel universes and time directionality. Neural network behavior in such a system would need to account for quantum states across dimensions while handling non-linear chaos effects that propagate both forward and backward through time across multiple universe states.
